File: 3355-zero-array-transformation-i/3355-zero-array-transformation-i.java
class Solution {
    public boolean isZeroArray(int[] nums, int[][] queries) {
        int n = nums.length;
        int[] p = new int[n+1];
        for(int[] q : queries) {
            p[q[0]] ++;
            p[q[1]+1] --;
        }
        long sum = 0;
        for(int i=0;i<n;i++) {
            sum  += p[i];
            if(sum<nums[i]) return false;
        }
        return true;
    }
}

File: 3355-zero-array-transformation-i/3355-zero-array-transformation-i.cpp
class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        vector<int> pref(n + 1);
        
        for (const auto& vec : queries) {
            int l = vec[0];
            int r = vec[1];
            
            pref[l]++;
            pref[r + 1]--;
        }
        
        for (int i = 1; i < n; i++) pref[i] += pref[i - 1];
        for (int i = 0; i < n; i++) if (pref[i] < nums[i]) return false;
        return true;
    }
};